src/device-state.coffee

```diff
@@ -328,61 +328,72 @@ module.exports = class DeviceState extends EventEmitter
 	loadTargetFromFile: (appsPath) ->
 		console.log('Attempting to load preloaded apps...')
 		appsPath ?= constants.appsJsonPath
-		fs.readFileAsync(appsPath, 'utf8')
-		.then(JSON.parse)
-		.then (stateFromFile) =>
-			if _.isArray(stateFromFile)
-				# This is a legacy apps.json
-				console.log('Legacy apps.json detected')
-				return @_convertLegacyAppsJson(stateFromFile)
-			else
-				return stateFromFile
-		.then (stateFromFile) =>
-			commitToPin = null
-			appToPin = null
-			if !_.isEmpty(stateFromFile)
-				images = _.flatMap stateFromFile.apps, (app, appId) =>
-					# multi-app warning!
-					# The following will need to be changed once running multiple applications is possible
-					commitToPin = app.commit
-					appToPin = appId
-					_.map app.services, (service, serviceId) =>
-						svc = {
-							imageName: service.image
-							serviceName: service.serviceName
-							imageId: service.imageId
-							serviceId
-							releaseId: app.releaseId
-							appId
-						}
-						return @applications.imageForService(svc)
-				Promise.map images, (img) =>
-					@applications.images.normalise(img.name)
-					.then (name) =>
-						img.name = name
-						@applications.images.save(img)
-				.then =>
-					@deviceConfig.getCurrent()
-				.then (deviceConf) =>
-					@deviceConfig.formatConfigKeys(stateFromFile.config)
-					.then (formattedConf) =>
-						stateFromFile.config = _.defaults(formattedConf, deviceConf)
-						stateFromFile.name ?= ''
-						@setTarget({
-							local: stateFromFile
-						})
-				.then =>
-					console.log('Preloading complete')
-					if stateFromFile.pinDevice
+		# Ensure that this is actually a file, and not an empty path
+		# It can be an empty path because if the file does not exist
+		# on host, the docker daemon creates an empty directory when
+		# the bind mount is added
+		fs.lstatAsync(appsPath)
+		.then (stats) ->
+			return stats.isFile()
+		.then (found) =>
```

Comment:
Checking if a file exists before opening it is not a good practice, it can lead to races (even if in this case it's unlikely).
We should `readFileAsync` directly and catch `ENOENT` and `EISDIR` instead.

Also I can't way for the day we use await / async in the supervisor.